### PR TITLE
wasm: refactor around stats, events and WasmExtension class.

### PR DESCRIPTION
**Pull request by mathetake**
This PR refactors the Wasm codebase around host defined stats, and WasmEvent, and WasmExtension class. Notablely, we delete the WasmExtension and EnvoyExtension classes which seem unused and unnecessary abstraction, and add CreateStatsHandler and LifecycleStatsHandler classes for handling Events and host defined metrics. The primary purpose is to drive the future work for the  https://github.com/envoyproxy/envoy/issues/16726 and https://github.com/envoyproxy/envoy/issues/16403. 

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Review by PiotrSikora — CHANGES_REQUESTED** [review]
This is actually used by Istio to export more stats. See:
* https://github.com/istio/proxy/pull/2916
* https://github.com/istio/proxy/blob/master/src/envoy/extensions/wasm/wasm.cc

Having said that, it seems that those stats all refer to async fetch, which we don't use anymore, so perhaps it could be removed... cc @mandarjog @kyessenov @bianpengyuan

**Comment by mathetake** [comment]
Thanks @PiotrSikora, I should have looked at istio/proxy, but yeah as you said, it looks like a redundant and not used anymore. 

Also I realized that we could remove [`IstioContext`](https://github.com/istio/proxy/blob/master/src/envoy/extensions/wasm/context.h#L23) there as well which I think is the reason for why factories are exposed to Istio via EnvoyWasm class and also why `class IstioWasm` exists there. 

I am more than happy to refactor istio/proxy as well since this could bring benefits both to Envoy and Istio/proxy codebase.

**Comment by mandarjog** [comment]
We can get rid of async fetch, it is not used anymore.
@yuval-k do you use async fetch ?

**Comment by PiotrSikora** [comment]
> We can get rid of async fetch, it is not used anymore.

That's a bit of a stretch. The fact that Istio doesn't use it anymore doesn't mean that other Envoy users are not using it.

In any case, if we want to have this discussion, could you please open a new issue? It's a breaking change that shouldn't be discussed in comments of a "random" PR.

**Comment by mandarjog** [comment]
@PiotrSikora agreed wrt Istio, so that cannot be the reason.

however the async fetch is racy. @kyessenov @bianpengyuan can we comment on why we decided that we cannot recommend async fetch alone?

File: source/extensions/common/wasm/stats_handler.cc [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  case WasmEvent::VMCreated:
+    runtime_stats_.active_.set(++active_wasms);
+    runtime_stats_.created_.inc();
+    break;
+  default:
+    break;
+  }
+}
+
+int64_t RuntimeStatsHandler::getActiveVMCount() { return active_wasms; };
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/stats_handler.cc:83]: Style: prefer camelCase (i.e. `VM` -> `Vm`). Same throughout rest of this PR.

**Review by PiotrSikora — COMMENTED** [review]
LGTM, sans nits.

**Comment by mathetake** [comment]
I found the coverage flakiness on Context::decodeData in context.cc: https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/wasm/context.cc#L1723

OK: https://storage.googleapis.com/envoy-pr/f4e1e1e/coverage/source/extensions/common/wasm/context.cc.gcov.html
Fail: https://storage.googleapis.com/envoy-pr/f413209/coverage/source/extensions/common/wasm/context.cc.gcov.html

retest for now.

**Comment by mathetake** [comment]
hmm.. looks like a deterministic but I have no idea why this affects these lines after merge main..

**Comment by PiotrSikora** [comment]
```
Code coverage for source/extensions/common is lower than limit of 96.5 (96.4)
Code coverage for source/extensions/common/wasm is lower than limit of 95.3 (95.2)
```
you need to adjust those values in `test/per_file_coverage.sh`, or ideally improve the coverage.

**Comment by mathetake** [comment]
yeah but I want to understand why the coverage drops after merging main: 

- Before main merge: https://storage.googleapis.com/envoy-pr/f4e1e1e/coverage/source/extensions/common/wasm/index.html
- After main merge: https://storage.googleapis.com/envoy-pr/f413209/coverage/source/extensions/common/wasm/index.html

I see that some lines of `convertFilterDataStatus`, `encodeData` and `decodeData` has started not being covered after main merge, which seems unrelated to this PR 😞

**Comment by mathetake** [comment]
OK https://github.com/envoyproxy/envoy/pull/17280/files#diff-a2dee6268db4920c2bd9665ba1429c144cd6a9e158b8845c503fb3d5c6fa273cL410-L411 this dropped the coverage on  encode/decodeData and convertFilterDataStatus, and this PR triggered the coverage CI failure somehow.

**Comment by mathetake** [comment]
raised the PR to backfill the coverage https://github.com/envoyproxy/envoy/pull/17292

**Comment by PiotrSikora** [comment]
> OK https://github.com/envoyproxy/envoy/pull/17280/files#diff-a2dee6268db4920c2bd9665ba1429c144cd6a9e158b8845c503fb3d5c6fa273cL410-L411 this dropped the coverage on encode/decodeData and convertFilterDataStatus, and this PR triggered the coverage CI failure somehow.

Oops, good catch! I wonder why the CI didn't catch it in that PR, though.

**Comment by kyessenov** [comment]
Async fetch has no error recovery. If the retries fail and xDS is ACKed, the error is permanent. So it is not recommended unless the network is reliable (which is never true).

**Review by PiotrSikora — APPROVED** [review]
Thanks!

**Comment by yuval-k** [comment]
> We can get rid of async fetch, it is not used anymore.
> @yuval-k do you use async fetch ?

we also are moving to ECDS, but are using async fetch in the interm.

**Pull request merged**